core/state_transition.go

```diff
@@ -220,80 +228,35 @@ func (st *StateTransition) TransitionDb() (ret []byte, usedGas uint64, failed bo
 	}
 
 	var (
-		evm = st.evm
 		// vm errors do not effect consensus and are therefore
 		// not assigned to err, except for insufficient balance
 		// error.
 		vmerr error
 	)
 
-	to := "<nil>"
-	if msg.To() != nil {
-		to = msg.To().Hex()
-	}
-
-	executionMgrTime := st.evm.Time
-	if executionMgrTime.Cmp(big.NewInt(0)) == 0 {
-		executionMgrTime = big.NewInt(1)
-	}
-
-	// TODO: queue origin is always 0 with current version of em
-	queueOrigin := big.NewInt(0)
-
-	l1MessageSender := msg.L1MessageSender()
-	if l1MessageSender == nil {
-		addr := common.HexToAddress("")
-		l1MessageSender = &addr
-	}
-
-	log.Debug("Applying transaction", "from", sender.Address().Hex(), "to", to, "nonce", msg.Nonce(), "l1MessageSender", l1MessageSender.Hex(), "data", hexutil.Encode(msg.Data()))
-
 	if contractCreation {
-		// Here we are going to call the EM directly
-		deployContractCalldata, _ := executionManagerAbi.Pack(
-			"executeTransaction",
-			executionMgrTime,        // lastL1Timestamp
-			queueOrigin,             // queueOrigin
-			common.HexToAddress(""), // ovmEntrypoint
-			st.data,                 // callBytes
-			sender,                  // fromAddress
-			l1MessageSender,         // l1MsgSenderAddress
-			true,                    // allowRevert
-		)
-
-		ret, st.gas, vmerr = evm.Call(sender, vm.ExecutionManagerAddress, deployContractCalldata, st.gas, st.value)
+		ret, _, st.gas, vmerr = st.evm.Create(sender, st.data, st.gas, st.value)
 	} else {
-		callContractCalldata, _ := executionManagerAbi.Pack(
-			"executeTransaction",
-			executionMgrTime, // lastL1Timestamp
-			queueOrigin,      // queueOrigin
-			st.to(),          // ovmEntrypoint
-			st.data,          // callBytes
-			sender,           // fromAddress
-			l1MessageSender,  // l1MsgSenderAddress
-			true,             // allowRevert
-		)
-
-		ret, st.gas, vmerr = evm.Call(sender, vm.ExecutionManagerAddress, callContractCalldata, st.gas, st.value)
+		// Increment the nonce for the next transaction
+		if !vm.UsingOVM {
+			st.state.SetNonce(msg.From(), st.state.GetNonce(msg.From())+1)
+		}
+		ret, st.gas, vmerr = st.evm.Call(sender, st.to(), st.data, st.gas, st.value)
 	}
-	if vmerr != nil {
-		log.Debug("VM returned with error", "err", vmerr)
-
-		// If the tx fails we won't have incremented the nonce. In this case, increment it manually
-		log.Debug("Incrementing nonce due to transaction failure")
-		st.state.SetNonce(msg.From(), st.state.GetNonce(sender.Address())+1)
```

Comment:
We need to figure out the best way to handle this nonce eviction thing. @karlfloersch should we just go ahead and update the eviction logic?